mathics/core/symbols.py

```diff
@@ -128,202 +128,34 @@ def __ne__(self, other) -> bool:
 
 class BaseExpression(KeyComparable):
     """
-    This is the base class from which all other Expressions are derived from.
+    This is the base class from which all other Expressions are
+    derived from.  If you think of an Expression as tree-like, then a
+    BaseExpression is a node in the tree.
 
     This class is not complete in of itself and subclasses should adapt or fill in
     what is needed
 
     Some important subclasses: Atom and Expression.
     """
 
-    options: Any
-    pattern_sequence: bool
-    unformatted: Any
     last_evaluated: Any
     # this variable holds a function defined in mathics.core.expression that creates an expression
     create_expression: Any
 
+    # __new__ seems to be used because this object references itself.
+    # In particular:
```

Comment:
@rocky, Notice that some descendants of BaseExpression decide the class of the object according to the input. For example, ``Real`` receives a float or a string and a precision parameter. According to the parameter, a ``MachineReal`` or a ``PrecisionReal`` object is created. In the case of ``Symbol``, we use ``__new__`` to return always the same instance of symbols with the same name.

Reply:
@mmatera I had considered that, but experiments show that this might not be the case. 

Consider this code: 

```
class Foo:
    def __new__(cls, *args, **kwargs):
        self = object.__new__(cls)
        print("XXX", id(self))
        return self


x = Foo()
y = Foo()

print(id(x))
print(id(y))
print(x == y)
```

If you run it you will see that `x` and `y` have different memory locations. And that x is not the same as y. 

So what leads you to believe that "new" is doing this? 

Also, I should say that if you wanted a single memory instance there is a design pattern to do this, but this isn't the way.

Reply:
@rocky, no: I am talking about ``Symbol``. I just mention this because I found that sometimes problems appear if you redefine ``__new__`` in a subclass with a different pattern of arguments than the default ``__new__`` in the parent class has.

Reply:
A link to a line(s) of code would be nice.

Reply:
Also, when there is non-obvious stuff going on, comments in code are a good place to describe this. If that gets too long an involved, we also have a developer's guide. 

Github comments in PRs will probably get lost. 

Think of coding like doing school homework in certain classes: you don't get credit for getting the right answer, you also need to show you work and calculcations.

Reply:
One other thing. I am drafting a discussion for this whole Expression mess. So let's hold off on detailed discussions until I post that, and we can carry on the discussions there.

Reply:
Here is my [small] opinion: maybe making every expression have a different memory allocation is *good* (makes the things "easier" in other places). But Python has exited in the last decades, if they have chosen to do it this way, probably there is a *why* (not saying that Python makes the best design choices, it doesn't). So we *should* (I'm not asking anyone to do this, just saying my opinion) do it the Pythonic way.